### PR TITLE
[InstCombine] When -A + B both have nsw flag, set nsw flag.

### LINKED ISSUE #72119: [InstCombine] Clang failed to fold (-a + b) / (a - b) into -1 due to the lack of nsw flag.
Alive2: https://alive2.llvm.org/ce/z/tPAt3K

godbolt: https://godbolt.org/z/1xhv9afKj

### PR DESCRIPTION
**Pull request by Z572**
Fixes #72119

https://alive2.llvm.org/ce/z/5f_QuC

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 1]
```diff
@@ -1487,7 +1487,11 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
       return BinaryOperator::CreateNeg(Builder.CreateAdd(A, B));
 
     // -A + B --> B - A
-    return BinaryOperator::CreateSub(RHS, A);
+    auto *Sub = BinaryOperator::CreateSub(RHS, A);
+    if (auto *Op1 = dyn_cast<BinaryOperator>(LHS)) {
+      Sub->setHasNoSignedWrap(I.hasNoSignedWrap() && Op1->hasNoSignedWrap());
+    }
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1493]:
```suggestion
    auto *OBO = cast<OverflowingBinaryOperator>(LHS));
    Sub->setHasNoSignedWrap(I.hasNoSignedWrap() && OBO->hasNoSignedWrap());
```

File: llvm/test/Transforms/InstCombine/add.ll [hunk 2]
```diff
@@ -120,6 +120,36 @@ define i32 @test5(i32 %A, i32 %B) {
   ret i32 %D
 }
 
+define i32 @test5_2(i32 %A, i32 %B) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/add.ll:123]:
```suggestion
define i32 @test5_both_nsw(i32 %A, i32 %B) {
```

File: llvm/test/Transforms/InstCombine/add.ll [hunk 3]
```diff
@@ -120,6 +120,36 @@ define i32 @test5(i32 %A, i32 %B) {
...
+; CHECK-LABEL: @test5_2(
+; CHECK-NEXT:    [[D:%.*]] = sub nsw i32 [[B:%.*]], [[A:%.*]]
+; CHECK-NEXT:    ret i32 [[D]]
+;
+  %C = sub nsw i32 0, %A
+  %D = add nsw i32 %C, %B
+  ret i32 %D
+}
+
+define i32 @test5_3(i32 %A, i32 %B) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/add.ll:133]:
```suggestion
define i32 @test5_neg_nsw(i32 %A, i32 %B) {
```

File: llvm/test/Transforms/InstCombine/add.ll [hunk 4]
```diff
@@ -120,6 +120,36 @@ define i32 @test5(i32 %A, i32 %B) {
...
+; CHECK-LABEL: @test5_3(
+; CHECK-NEXT:    [[D:%.*]] = sub i32 [[B:%.*]], [[A:%.*]]
+; CHECK-NEXT:    ret i32 [[D]]
+;
+  %C = sub nsw i32 0, %A
+  %D = add i32 %C, %B
+  ret i32 %D
+}
+
+define i32 @test5_4(i32 %A, i32 %B) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/add.ll:143]:
```suggestion
define i32 @test5_add_nsw(i32 %A, i32 %B) {
```

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 5]
```diff
@@ -1487,7 +1487,11 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
       return BinaryOperator::CreateNeg(Builder.CreateAdd(A, B));
 
     // -A + B --> B - A
-    return BinaryOperator::CreateSub(RHS, A);
+    auto *Sub = BinaryOperator::CreateSub(RHS, A);
+    auto *OB0 = dyn_cast<OverflowingBinaryOperator>(LHS);
+    Sub->setHasNoSignedWrap(I.hasNoSignedWrap() && OB0->hasNoSignedWrap());
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1492]: Based on your proof think all you need is the `I.hasNoSignedWrap()`

Comment (Z572) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1492]: https://alive2.llvm.org/ce/z/7nnL25

if only `I.hasNoSignedWrap()`

```
define i32 @test5_add_nsw(i32 %A, i32 %B) {
  %C = sub i32 0, %A
  %D = add nsw i32 %C, %B
  ret i32 %D
}
```
--> 
```
define i32 @test5_add_nsw(i32 %A, i32 %B) {
  %D = sub nsw i32 %B, %A
  ret i32 %D
}
```

is error

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1492]: yeah, sorry I misread.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 1 above)
Comment (Z572) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: https://github.com/llvm/llvm-project/pull/72282

**Pull request merged**